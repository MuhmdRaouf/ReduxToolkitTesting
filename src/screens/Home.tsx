import type { RootState } from '../types/redux';
import { Button, Layout, Spinner, Text } from '@ui-kitten/components';
import React, { useCallback, useEffect } from 'react';
import { Dimensions, Image, StyleSheet, View } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';

import { fetchPicture, deletePicture } from '../redux/gallery/actions';

export const { width } = Dimensions.get('window');

export const Home = () => {
  const dispatch = useDispatch();
  const defaultPicture =
    'https://akveo.github.io/react-native-ui-kitten/docs/assets/playground-build/static/media/icon.a78e4b51.png';
  const { picture, isFetching, fetchErrorMessage } = useSelector((state: RootState) => state.gallery);

  const fetchPictureHandle = useCallback(() => {
    const pictureRepo = 'https://picsum.photos/v2/list';
    dispatch(fetchPicture({ pictureRepo }));
  }, [dispatch]);

  const ResetPictureHandle = useCallback(() => {
    dispatch(deletePicture(defaultPicture));
  }, [dispatch]);

  const brokenFetchPictureHandle = useCallback(() => {
    const pictureRepo = 'https://picsum.photos/v2/broke-link';
    dispatch(fetchPicture({ pictureRepo }));
  }, [dispatch]);

  useEffect(() => {
    fetchPictureHandle();
  }, [fetchPictureHandle]);

  return (
    <Layout style={styles.container}>
      <View style={styles.pictureContainer}>
        {isFetching ? <Spinner size="giant" /> : <Image style={styles.picture} source={{ uri: picture }} />}
      </View>
      {fetchErrorMessage && (
        <Text status="danger" category="h5">
          {fetchErrorMessage}
        </Text>
      )}
      <View style={styles.buttonContainer}>
        <Button style={styles.button} onPress={fetchPictureHandle}>
          <Text>Fetch new picture</Text>
        </Button>
        <Button style={styles.button} onPress={ResetPictureHandle}>
          <Text>Reset Picture</Text>
        </Button>
        <Button style={styles.button} onPress={brokenFetchPictureHandle}>
          <Text>Broken Button</Text>
        </Button>
      </View>
    </Layout>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  pictureContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  picture: {
    width: width * 0.8,
    height: width * 0.8,
    margin: 10,
    borderRadius: 10,
  },
  buttonContainer: {
    width: '100%',
    justifyContent: 'space-evenly',
    alignItems: 'center',
    marginBottom: 25,
  },
  button: {
    width: width / 2,
    marginBottom: 10,
  },
});
