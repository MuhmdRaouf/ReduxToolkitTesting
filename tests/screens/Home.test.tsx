/* eslint-disable camelcase */
import { StatusCodes } from 'http-status-codes';
import React from 'react';

import { Home } from '../../src/screens/Home';
import { axiosMock } from '../mockedAxios';
import { render, cleanup } from '../testUtils';

describe('<Home />', () => {
  beforeEach(() => {
    axiosMock.reset();
    jest.restoreAllMocks();
  });

  afterEach(() => {
    jest.resetAllMocks();
    cleanup();
  });

  it('should match <Home /> snapshot', () => {
    expect.hasAssertions();

    const { toJSON } = render(<Home />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should fire fetchPicture action on mount', () => {
    expect.hasAssertions();

    const pictureRepo = 'https://picsum.photos/v2/list';
    axiosMock.onGet(pictureRepo).reply(StatusCodes.OK, [{ download_url: pictureRepo }]);

    const { store } = render(<Home />);

    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'FETCH_PICTURE/fulfilled',
      payload: { picture: pictureRepo },
    });
  });
});
