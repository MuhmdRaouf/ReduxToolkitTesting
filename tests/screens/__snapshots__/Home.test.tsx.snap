// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should match <Home /> snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    appearance="default"
    level="1"
    style={
      Array [
        Object {
          "backgroundColor": "#222B45",
        },
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Size {
            "height": 32,
            "width": 32,
          }
        }
      >
        <View
          style={
            Object {
              "bottom": 0,
              "height": 32,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 32,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                undefined,
                Size {
                  "height": 16,
                  "width": 32,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "height": 32,
                  "transform": Array [
                    Object {
                      "rotate": "-165deg",
                    },
                  ],
                  "width": 32,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Size {
                      "height": 16,
                      "width": 32,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderColor": "#3366FF",
                        "borderRadius": 16,
                        "borderWidth": 4.5,
                        "height": 32,
                        "width": 32,
                      },
                      undefined,
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "bottom": 0,
              "height": 32,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 32,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "top": 16,
                },
                Size {
                  "height": 16,
                  "width": 32,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "height": 32,
                  "transform": Array [
                    Object {
                      "translateY": -16,
                    },
                    Object {
                      "rotate": "345deg",
                    },
                  ],
                  "width": 32,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Size {
                      "height": 16,
                      "width": 32,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderColor": "#3366FF",
                        "borderRadius": 16,
                        "borderWidth": 4.5,
                        "height": 32,
                        "width": 32,
                      },
                      undefined,
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "space-evenly",
          "marginBottom": 25,
          "width": "100%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3366FF",
            "borderColor": "#3366FF",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "minHeight": 40,
            "minWidth": 40,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 12,
            "width": 375,
          }
        }
      >
        <Text
          appearance="default"
          category="p1"
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "bold",
                "marginHorizontal": 10,
              },
            ]
          }
        >
          Fetch new picture
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3366FF",
            "borderColor": "#3366FF",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "minHeight": 40,
            "minWidth": 40,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 12,
            "width": 375,
          }
        }
      >
        <Text
          appearance="default"
          category="p1"
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "bold",
                "marginHorizontal": 10,
              },
            ]
          }
        >
          Reset Picture
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3366FF",
            "borderColor": "#3366FF",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "minHeight": 40,
            "minWidth": 40,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 12,
            "width": 375,
          }
        }
      >
        <Text
          appearance="default"
          category="p1"
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "bold",
                "marginHorizontal": 10,
              },
            ]
          }
        >
          Broken Button
        </Text>
      </View>
    </View>
  </View>
</View>
`;
